lms/djangoapps/completion/waffle.py

```diff
@@ -0,0 +1,20 @@
+"""
+This module contains various configuration settings via
+waffle switches for the completion app.
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from openedx.core.djangoapps.waffle_utils import WaffleSwitchNamespace
+
+# Namespace
+WAFFLE_NAMESPACE = 'completion'
+
+# Switches
+ENABLE_COMPLETION_TRACKING = 'enable_completion_tracking'
```

Comment:
@jcdyer: The old Mercury team tended to use Waffle Flags (rather than switches) for everything.  I like that the flags enable you to turn things on for single users, or beta-testers, or an individual course (see CourseWaffleFlag), that isn't available with switches.  Also, I think our waffle_utils code for flags is a bit cleaner than with switches.

Would you mind reviewing some of the flags here and let let me know what you think?
https://github.com/edx/edx-platform/blob/5ae2bee17f77e345fadef3a0935dc85a890ec405/openedx/features/course_experience/__init__.py

Reply:
Unless I misunderstand something, waffle flags require a request object to be present.  A lot of this code is happening in signals and on models.  Would you recommend refactoring to block things at the view level or passing a fake request object with the relevant user attached to it?

Reply:
We had debated using `RequestCache.get_current_request()` internally and taking `request` out of the signature, but wanted to continue to make this explicit.

Would it work to use `RequestCache.get_current_request()`?  I've seen a couple of other examples of using the request cache in models.

Reply:
Unless you think it's very important, I'd rather not do that.  Another client had me spend the last several weeks removing code that was doing something similar in a different codebase, because it was making refactoring and testing a nightmare.  There isn't always a current request available (tests, celery tasks, and management commands are three examples), and then to use any of this code in any of those contexts, a fake request needs to be set up by the caller and injected into the request cache.  Worse, the fact that this is needed is not evident from the signatures of the methods that check the waffle flag, so you can't work with the code without intimate knowledge of its internals.  

If we have enough information available inside all the contexts where the flag is checked to be able to create a fake response inside the public interfaces, then it would make sense.  I haven't looked closely enough at waffle to know exactly what it needs from a request.  I know we'll have a user and a course available, because it's already part of the submit_completion signature.  But I don't know if that's enough, or if not, how much extra work it will be to create a suitable request.

A waffle switch gives us the simple ability to turn the feature on and off without the extra complication, and my understanding was that this was all we needed for this--to be sure we don't start saving data until the feature is complete (or complete enough).

Reply:
Got it.  A switch covers the most important need for now, and hopefully the only need. :)